### PR TITLE
HDDS-4123. Integrate OM Open Key Cleanup Service Into Existing Code

### PR DESCRIPTION
**Pull request by kaijchen**
## What changes were proposed in this pull request?

~~WIP: this PR depends on #3233 and #3226.~~

This pull request completes the open key cleanup service outlined in the parent Jira HDDS-4120. It implements the `OpenKeyCleanupService` class, and starts and stops the service in `KeyManagerImpl`. The following configurations have been defined to specify the service's behavior:

1. ozone.open.key.cleanup.service.interval (default value 24 hours)
2. ozone.open.key.expire.threshold (default value 1 week)
3. ozone.open.key.cleanup.limit.per.task (default value 1000 keys)

See `ozone-defaults.xml` for their corresponding descriptions.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4123

## How was this patch tested?

Integration test `TestOpenKeyCleanupService`

### CODE REVIEW

**Comment by kaijchen** [comment]
@errose28 @captainzmc PTAL.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOpenKeyCleanupService.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+  /**
+   * In this test, we create a bunch of keys and delete them. Then we start the
+   * KeyDeletingService and pass a SCMClient which does not fail. We make sure
+   * that all the keys that we deleted is picked up and deleted by
+   * OzoneManager.
+   *
+   * @throws IOException - on Failure.
+   */
+
```
Comment (kaijchen) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOpenKeyCleanupService.java:116]:
```suggestion
```

**Comment by kaijchen** [comment]
100x TestOpenKeyCleanupService: https://github.com/kaijchen/ozone/actions/runs/2452367006

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java [hunk 2]
```diff
@@ -85,8 +83,6 @@ public void init() throws Exception {
     omId = UUID.randomUUID().toString();
     conf.setBoolean(OZONE_ACL_ENABLED, true);
     conf.setBoolean(OZONE_SECURITY_ENABLED_KEY, true);
-    conf.setTimeDuration(OZONE_OM_OPEN_KEY_EXPIRE_THRESHOLD,
-        2, TimeUnit.SECONDS);
```
Comment (captainzmc) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java:89]: Does the OpenKey Clean Update Service cause this CI to fail?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 3]
```diff
@@ -73,18 +165,72 @@ public int getPriority() {
...
+              .setClientId(clientId)
+              .setServerId(server.getRaftPeerId())
+              .setGroupId(server.getRaftGroupId())
+              .setCallId(runCount.get())
+              .setMessage(Message.valueOf(
+                  OMRatisHelper.convertRequestToByteString(omRequest)))
+              .setType(RaftClientRequest.writeRequestType())
+              .build();
+
+          server.submitRequest(omRequest, raftClientRequest);
```
Comment (captainzmc) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java:227]: After successful deletion.  Can we add some metrics to show how many open keys has been deleted? This should be helpful in production.

**Review by captainzmc — COMMENTED** [review]
Thanks @kaijchen  for the working, LGTM.  Just have some minor comments here.

**Comment by captainzmc** [comment]
Hi @errose28, this task is the last part of HDDS-4120, through which open key can be cleaned normally. Could you also help review this?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java (hunk 2 above)
Comment (kaijchen) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java:89]: I don't think so. IIRC, this test is flaky before.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java (hunk 3 above)
Comment (kaijchen) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java:227]: I think metrics is already included in OMOpenKeysDeleteRequest.

https://github.com/apache/ozone/blob/377be9644fb2e7a1b6b6413a82ac15e3dc4c48d3/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMOpenKeysDeleteRequest.java#L67

https://github.com/apache/ozone/blob/377be9644fb2e7a1b6b6413a82ac15e3dc4c48d3/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMOpenKeysDeleteRequest.java#L81

https://github.com/apache/ozone/blob/377be9644fb2e7a1b6b6413a82ac15e3dc4c48d3/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMOpenKeysDeleteRequest.java#L130

**Comment by kaijchen** [comment]
Thanks @captainzmc for the review. I think there is one more task left after this, HDDS-6769.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java (hunk 2 above)
Comment (kaijchen) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSecureOzoneManager.java:89]: There is no problem in TestSecureOzoneManager with or without HDDS-4123

https://github.com/kaijchen/ozone/actions/runs/2454403397
https://github.com/kaijchen/ozone/actions/runs/2454495538

**Review by captainzmc — APPROVED** [review]

**Comment by captainzmc** [comment]
Hi @kaijchen , did you tested this patch in local cluster? For example, keep writing failed keys. The clean up service keep deleting in the background. Run a few days for stability test to see if there's anything problem.

**Comment by kaijchen** [comment]
Yes, I have tested it in a local cluster. From the log we can see open key cleanup service works as expected.
The loglevel was changed to `INFO` for testing purpose.

```
2022-06-17 15:52:14,891 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-17 15:52:14,892 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-17 16:02:14,891 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-17 16:02:14,922 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: Number of expired keys submitted for deletion: 20, elapsed time: 31ms
2022-06-17 16:02:14,922 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-17 16:12:14,891 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-17 16:12:14,891 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
```

**Comment by kaijchen** [comment]
Limiting number of expired keys is working as expected.

```
2022-06-18 22:42:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 22:42:14,976 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 22:52:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 22:52:14,990 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: Number of expired keys submitted for deletion: 1000, elapsed time: 80ms
2022-06-18 22:52:14,990 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:02:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:02:14,995 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: Number of expired keys submitted for deletion: 1000, elapsed time: 86ms
2022-06-18 23:02:14,996 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:12:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:12:14,979 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: Number of expired keys submitted for deletion: 108, elapsed time: 69ms
2022-06-18 23:12:14,979 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:22:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:22:14,965 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
2022-06-18 23:32:14,910 [OpenKeyCleanupService#0] INFO org.apache.hadoop.ozone.om.OpenKeyCleanupService: running open key cleanup task
```

**Comment by captainzmc** [comment]
> Limiting number of expired keys is working as expected.
> 
The tests looked good, thanks @kaijchen working on this.  Let's merge this PR first, if other modifications are needed, we can submit a new patch.

**Comment by kaijchen** [comment]
Thanks @captainzmc for reviewing and merging this.

**Pull request merged**